dialogue-target/src/main/java/com/palantir/dialogue/Request.java

```diff
@@ -16,37 +16,203 @@
 
 package com.palantir.dialogue;
 
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableMultimap;
+import com.google.common.collect.ImmutableSortedMap;
 import com.google.common.collect.Multimap;
+import com.palantir.logsafe.Preconditions;
 import java.util.Map;
+import java.util.Objects;
 import java.util.Optional;
-import org.immutables.value.Value;
 
 /** Defines the parameters of a single {@link Call} to an {@link Endpoint}. */
-@DialogueImmutablesStyle
-@Value.Immutable
-public interface Request {
-    /** The HTTP headers for this request, encoded as a map of {@code header-name: header-value}. */
-    Map<String, String> headerParams();
```

Comment:
perhaps a slightly smaller change and one with less hand-written code would be changing the build method of the builder class to make a TreeMap and then copy it back into the object, something like

```java
Request request = super.build();
Map<String, String> caseInsensitiveHeaderParams = new TreeMap<>(Case...);
caseInsensitiveHeaderParams.putAll(request.headerParams());
return request.toBuilder().headerParams(caseInsensitiveHeaderParams).build();
```

Reply:
or alternatively, we could replace the builder methods for adding headerParams to run toLowerCase on the strings at build time?

Reply:
one more alternative: we could leave this interface as-is and manually write the implementation class + Builder and return the implemented interface vs. making the entire thing concrete

Reply:
The problem is consumers writing code along the lines of `request.headerParams().containsKey("Content-Type")` would fail when `"content-type"` is present. I'm worried that the asymmetry between Request and Response headers will make it easy to introduce subtle bugs that aren't easily caught in review.

Reply:
>  we could leave this interface as-is and manually write the implementation class + Builder

That's true, but is it desirable in this case? Do we want to allow custom Request implementations? We could avoid the break, but I think now is the time to make breaks before we roll this out more broadly.